music/src/core/data.ts

```diff
@@ -493,3 +501,242 @@ export class TrioConverter extends DataConverter {
     return ns;
   }
 }
+
+/**
+ * Converts between a quantized multitrack `NoteSequence` and `Tensor` objects
+ * used by `MusicVAE`.
+ *
+ * Each track is represented using events from the following vocabulary:
+ *   - An initial program-select event specifying which MIDI program to use.
+ *   - A sequence of performance (note-on, note-off, time-shift,
+ *     velocity-change) events.
+ *   - An end token.
+ *
+ * Tracks are ordered by program number with drums at the end, then one-hot
+ * encoded and padded with zeros to the maximum number of events. If fewer than
+ * the maximum number of tracks are present, extra tracks consisting of only an
+ * end token (then one-hot encoded and zero-padded) will be added.
+ *
+ * @param numSteps The total number of events used to encode each
+ * `NoteSequence`.
+ * @param numSegments The number of tracks to use. `NoteSequence`s with more
+ * tracks will have tracks removed. `NoteSequence`s with fewer tracks will be
+ * padded with empty tracks.
+ * @param stepsPerQuarter The number of time steps per quarter note.
+ * @param totalSteps The length of each `NoteSequence` in time steps. Longer
+ * `NoteSequence`s will be truncated. Shorter `NoteSequence`s will be padded
+ * with silence.
+ * @param numVelocityBins The number of bins into which to quantize note
+ * velocities.
+ */
+export interface MultitrackConverterArgs extends BaseConverterArgs {
+  stepsPerQuarter: number;
+  totalSteps: number;
+  numVelocityBins: number;
+}
+export class MultitrackConverter extends DataConverter {
+  readonly NUM_SPLITS = 0;
+
+  readonly stepsPerQuarter: number;
+  readonly totalSteps: number;
+  readonly numVelocityBins: number;
+
+  readonly numPitches: number;
+  readonly performanceEventDepth: number;
+  readonly numPrograms: number;
+  readonly depth: number;
+
+  constructor(args: MultitrackConverterArgs) {
+    super(args);
+
+    this.stepsPerQuarter = args.stepsPerQuarter;
+    this.totalSteps = args.totalSteps;
+    this.numVelocityBins = args.numVelocityBins;
+
+    // Vocabulary:
+    // note-on, note-off, time-shift, velocity-change, program-select, end-token
+
+    this.numPitches = constants.MAX_MIDI_PITCH - constants.MIN_MIDI_PITCH + 1;
+    this.performanceEventDepth =
+        2 * this.numPitches + this.totalSteps + this.numVelocityBins;
+
+    // Include an extra "program" for drums.
+    this.numPrograms =
+        constants.MAX_MIDI_PROGRAM - constants.MIN_MIDI_PROGRAM + 2;
+    this.depth = this.performanceEventDepth + this.numPrograms + 1;
```

Comment:
maybe add a comment that the 1 is for end token

Reply:
Used explicit endToken property.